source/slang/hlsl.meta.slang

```diff
@@ -1076,14 +1076,34 @@ bool all(matrix<T,N,M> x)
 }
 
 // Barrier for writes to all memory spaces (HLSL SM 5.0)
-__target_intrinsic(glsl, "memoryBarrier(), groupMemoryBarrier(), memoryBarrierImage(), memoryBarrierBuffer()")
-__target_intrinsic(cuda, "__threadfence()")
-void AllMemoryBarrier();
+void AllMemoryBarrier()
+{
+    __target_switch
+    {
+    case hlsl: __intrinsic_asm "AllMemoryBarrier";
+    case glsl: __intrinsic_asm "memoryBarrier()";
+    case cuda: __intrinsic_asm "__threadfence()";
+    case spirv: spirv_asm
+        {
+            OpMemoryBarrier Device AcquireRelease|UniformMemory|WorkgroupMemory|ImageMemory;
+        };
+    }
+}
 
 // Thread-group sync and barrier for writes to all memory spaces (HLSL SM 5.0)
-__target_intrinsic(glsl, "memoryBarrier(), groupMemoryBarrier(), memoryBarrierImage(), memoryBarrierBuffer(), barrier()")
-__target_intrinsic(cuda, "__syncthreads()")
-void AllMemoryBarrierWithGroupSync();
+void AllMemoryBarrierWithGroupSync()
+{
+    __target_switch
+    {
+    case hlsl: __intrinsic_asm "AllMemoryBarrierWithGroupSync";
+    case glsl: __intrinsic_asm "barrier()";
```

Comment:
I don't think this is right. `barrier()` corresponds to `GroupMemoryBarrierWithGroupSync()`, but not to `AllMemoryBarrierWithGroupSync()`

https://github.com/KhronosGroup/GLSL/blob/ad383324612d8493c78d52d0c903969fe1dddd0a/extensions/khr/GL_KHR_memory_scope_semantics.txt#L461-L465
https://github.com/KhronosGroup/GLSL/blob/ad383324612d8493c78d52d0c903969fe1dddd0a/extensions/khr/GL_KHR_vulkan_glsl.txt#L678-L681

I don't think GLSL has a single built in function to cover this case like HLSL does. I'd recommend just using GL_KHR_memory_scope_semantics explicitly. The list of semantics would be very close to the SPIR-V version below as well, except the storage semantics don't map exactly from what I can tell.

Reply:
Fixed and verified that the slang->glslang->spirv output is the same with dxc's output.